### PR TITLE
lib/timeutils: parse_timestamp: bugfix, unittests and usec support

### PR DESCRIPTION
**Pull request by t-8ch**
Edit: also contains some changes for the shellcheck lint.

### CODE REVIEW

File: lib/timeutils.c [hunk 1]
```diff
@@ -147,7 +147,27 @@ static int parse_sec(const char *t, usec_t *usec)
...
+	while (*(++t)) {
+		if (!isdigit(*t) || factor < 1)
+			return 0;
+
+		ret += ((usec_t) *t - '0') * factor;
+		factor /= 10;
+	}
+
+	*usec = ret;
+	return 1;
```
Comment (karelzak) [inline lib/timeutils.c:167]: Would it be possible to return 0 on success and <0 on error? ... as we usually use ;-)

File: lib/timeutils.c [hunk 2]
```diff
@@ -147,7 +147,27 @@ static int parse_sec(const char *t, usec_t *usec)
...
+}
+
+static int parse_timestamp_reference(time_t x, const char *t, usec_t *usec)
 {
```
Comment (karelzak) [inline lib/timeutils.c:171]: There is `Allowed syntaxes` comment in the function. Please, add sub-seconds there.

**Comment by karelzak** [comment]
Cool, thanks for the test!

**Comment by karelzak** [comment]
The new regression test failed on my machine. 

The timeutils-timestamp.err diff (output improved by patch below):
```diff
+#00 2012-09-22 16:34:22      : 1348324462000000 != 1348331662000000
+#01 2012-09-22 16:34:22,012  : 1348324462012000 != 1348331662012000
+#02 2012-09-22 16:34:22.012  : 1348324462012000 != 1348331662012000
+#05 2012-09-22 16:34         : 1348324440000000 != 1348331640000000
+#06 2012-09-22               : 1348264800000000 != 1348272000000000
+#07 16:34:22                 : 1674228862000000 != 1674232462000000
+#08 16:34:22,123456          : 1674228862123456 != 1674232462123456
+#09 16:34:22.123456          : 1674228862123456 != 1674232462123456
+#10 16:34                    : 1674228840000000 != 1674232440000000
+#12 yesterday                : 1674082800000000 != 1674086400000000
+#13 today                    : 1674169200000000 != 1674172800000000
+#14 tomorrow                 : 1674255600000000 != 1674259200000000
+#17 20120922163422           : 1348324462000000 != 1348331662000000
```
See, for example, `today` where the diff is 3600 seconds  (1674169200000000 - 1674172800000000)/1000000. And for the first line (#00) the diff is 7200 seconds.

The patch to get more rich output from the test:
```diff
diff --git a/lib/timeutils.c b/lib/timeutils.c
index 8f23f04e2..7468f3197 100644
--- a/lib/timeutils.c
+++ b/lib/timeutils.c
@@ -620,6 +620,7 @@ time_t timegm(struct tm *tm)
 
 static int run_unittest_timestamp(void)
 {
+       int rc = EXIT_SUCCESS;
        time_t reference = 1674180427;
        static const struct testcase {
                const char * const input;
@@ -655,13 +656,13 @@ static int run_unittest_timestamp(void)
                }
 
                if (result != t.expected) {
-                       fprintf(stderr, "%s: %"PRId64" != %"PRId64"\n",
-                               t.input, result, t.expected);
-                       return EXIT_FAILURE;
+                       fprintf(stderr, "#%02zu %-25s: %"PRId64" != %"PRId64"\n",
+                               i, t.input, result, t.expected);
+                       rc = EXIT_FAILURE;
                }
        }
 
-       return 0;
+       return rc;
 }
 
 int main(int argc, char *argv[])
```

**Comment by t-8ch** [comment]
I integrated your patch, thanks.

The problem was your systems TZ ENV variable.
The unittest now unsets this variable.

**Comment by karelzak** [comment]
I have set TZ=GMT in the test.

**Pull request merged**